### PR TITLE
feat: Amazon EKS MCP Server

### PR DESCRIPTION
**Pull request by patrick-yu-amzn**
Fixes

### Changes

The Amazon EKS MCP server provides AI code assistants with resource management tools and real-time cluster state visibility. This provides  large language models (LLMs) with essential tooling and contextual awareness, enabling AI code assistants to streamline application development through tailored guidance—from initial setup through production optimization and troubleshooting.

Integrating the EKS MCP server into AI code assistants enhances development workflow across all phases, from simplifying initial cluster setup with automated prerequisite creation and application of best practices. Further, it streamlines application deployment with high-level workflows and automated code generation. Finally, it accelerates troubleshooting through intelligent debugging tools and knowledge base access. All of this simplifies complex operations through natural language interactions in AI code assistants.

### User experience

New capabilities added:

* Enables users of AI code assistants to create new EKS clusters, complete with prerequisites such as dedicated VPCs, networking, and EKS Auto Mode node pools, by translating requests into the appropriate AWS CloudFormation actions.
* Provides the ability to deploy containerized applications by applying existing Kubernetes YAML files or by generating new deployment and service manifests based on user-provided parameters.
* Supports full lifecycle management of individual Kubernetes resources (such as Pods, Services, and Deployments) within EKS clusters, enabling create, read, update, patch, and delete operations.
* Provides the ability to list Kubernetes resources with filtering by namespace, labels, and fields, simplifying the process for both users and LLMs to gather information about the state of Kubernetes applications and EKS infrastructure.
* Facilitates operational tasks such as retrieving logs from specific pods and containers or fetching Kubernetes events related to particular resources, supporting troubleshooting and monitoring for both direct users and AI-driven workflows.
* Enables users to troubleshoot issues with an EKS cluster.

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] I have reviewed the [contributing guidelines](https://github.com/awslabs/mcp/blob/main/CONTRIBUTING.md)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented

Is this a breaking change? (Y/N) N

**RFC issue number**: https://github.com/awslabs/mcp/issues/305

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the [project license](https://github.com/awslabs/mcp/blob/main/LICENSE).

### CODE REVIEW

File: src/eks-mcp-server/awslabs/eks_mcp_server/eks_kb_handler.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+        This tool provides troubleshooting guidance for Amazon EKS issues by querying
+        a specialized knowledge base of EKS troubleshooting information. It helps identify
+        common problems and provides step-by-step solutions for resolving cluster creation issues,
+        node group management problems, workload deployment issues, and diagnosing error messages.
+
+        ## Requirements
+        - Internet connectivity to access the EKS Knowledge Base API
+        - Valid AWS credentials with permissions to access the EKS Knowledge Base
+        - IAM permission: eks-mcpserver:QueryKnowledgeBase
```
Comment (SaranBalaji90) [inline src/eks-mcp-server/awslabs/eks_mcp_server/eks_kb_handler.py:60]: /s/Query/Get/

File: src/eks-mcp-server/README.md [hunk 2]
```diff
@@ -0,0 +1,464 @@
...
+  "mcpServers": {
+    "awslabs.eks-mcp-server": {
+      "command": "uv",
+      "args": [
+        "--directory",
+        "/path/to/awslabs/eks_mcp_server",
+        "run",
+        "server.py",
+        "--allow-write", // Optional parameter if you would like to allow MCP to take mutating actions (i.e. Create/Update/Delete Resource)
+        "--allow-sensitive-data-access" // Optional parameter if you would like to allow MCP to access sensitive data such as logs, events, and Kubernetes Secrets
```
Comment (csantanapr) [inline src/eks-mcp-server/README.md]:
```suggestion
        "--allow-write",
        "--allow-sensitive-data-access"
```
Remove comments from json code snippet, as user would copy and paste the code and this will cause errors in tools reading the json config, as the comments make the json section invalid

**Review by csantanapr — CHANGES_REQUESTED** [review]

**Review by awslabs-mcp — CHANGES_REQUESTED** [review]
## Pull Request Review - Amazon EKS MCP Server

I have conducted a comprehensive review of this pull request following the established review guidelines. This is a substantial contribution that adds EKS cluster management capabilities to the MCP ecosystem.

### Overall Assessment

This PR introduces a well-architected Amazon EKS MCP Server with comprehensive functionality for cluster management, Kubernetes resource operations, and CloudWatch integration. The implementation follows established patterns from other MCP servers in this repository and demonstrates good separation of concerns.

### Issues Requiring Resolution

#### 1. JSON Configuration Error (CRITICAL)
**Reference**: Review comment by csantanapr

The README.md contains JSON configuration examples with JavaScript-style comments which will cause JSON parsing errors when users copy-paste the configuration. This must be fixed before merging.

**Location**: src/eks-mcp-server/README.md:100
**Solution**: Remove comments from JSON snippets or move them outside the code blocks.

#### 2. Failed Required Check

The fail-by-label workflow failed, which is a required check for merging. This needs to be resolved before the PR can be approved.

**Workflow Run**: 15305242187

#### 3. Test Coverage Below Standards

While the overall coverage is good (84.22%), several core components have coverage below the repository standards:

- eks_stack_handler.py: 73.33% (48 lines uncovered)
- cloudwatch_handler.py: 71.51% (47 lines uncovered)  
- k8s_apis.py: 75.26% (47 lines uncovered)
- iam_handler.py: 78.75% (17 lines uncovered)

**Recommendation**: Add tests for error handling paths and edge cases in these core components.

### Positive Aspects

- Security Compliance: All security scans passed (Bandit, CodeQL, Semgrep, etc.)
- Architecture: Clean separation of concerns with dedicated handlers
- Documentation: Comprehensive README with detailed setup instructions
- Testing: Extensive test suite covering major functionality
- Standards Compliance: Follows established MCP server patterns
- Infrastructure: Proper use of CloudFormation for reliable deployments

### Action Items

1. CRITICAL: Fix JSON configuration examples in README.md (remove comments)
2. REQUIRED: Resolve failed fail-by-label workflow check
3. RECOMMENDED: Improve test coverage for core handler components
4. RECOMMENDED: Address any remaining Codecov feedback

### Conclusion

This is a high-quality contribution that will add significant value to the MCP ecosystem. The implementation is well-structured and follows best practices. Once the critical JSON configuration issue and required workflow check are resolved, this PR will be ready for approval.

**Decision**: REQUEST_CHANGES

Reference workflow run: 15305242051

Generated with Claude Code

Co-Authored-By: Claude <noreply@anthropic.com>

File: src/eks-mcp-server/README.md (hunk 2 above)
Comment (patrick-yu-amzn) [reply src/eks-mcp-server/README.md]: Already removed in the latest version of the README.md

File: src/eks-mcp-server/awslabs/eks_mcp_server/aws_helper.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        """Get the AWS region from the environment if set."""
+        return os.environ.get('AWS_REGION')
+
+    @staticmethod
+    def get_aws_profile() -> Optional[str]:
+        """Get the AWS profile from the environment if set."""
+        return os.environ.get('AWS_PROFILE')
+
+    @classmethod
+    def create_boto3_client(cls, service_name: str, region_name: Optional[str] = None) -> Any:
```
Comment (scottschreckengaust) [inline src/eks-mcp-server/awslabs/eks_mcp_server/aws_helper.py:38]: recommend adding:

```python
from botocore.config import Config
Config(
            user_agent_extra=f'awslabs/mcp/eks-mcp-server/{__version__}'
        )
# add to the session...
```

**Review by scottschreckengaust — APPROVED** [review]
Fast follow with version in user agent

**Review by awslabs-mcp — CHANGES_REQUESTED** [review]
# Pull Request Review - Amazon EKS MCP Server (Updated Assessment)

I have conducted a follow-up review of PR #429 following the comprehensive review process guidelines. This pull request introduces a substantial new Amazon EKS MCP Server with extensive functionality for cluster management and Kubernetes operations.

## Summary of Assessment

This is a well-architected contribution that follows established MCP server patterns and provides comprehensive EKS cluster management capabilities. However, there are several issues that need to be addressed before this PR can be approved.

## Key Issues Requiring Resolution

### 1. JSON Configuration Issue (CRITICAL - ADDRESSED)
The JSON configuration examples mentioned in the previous review by @csantanapr appear to have been resolved. The JSON snippets in the current diff are properly formatted without JavaScript-style comments.

### 2. Test Coverage Below Repository Standards (HIGH PRIORITY)
Multiple core components have test coverage below acceptable levels:

- eks_stack_handler.py: 73.33% (48 lines uncovered)
- cloudwatch_handler.py: 71.51% (47 lines uncovered)  
- k8s_apis.py: 75.26% (47 lines uncovered)
- iam_handler.py: 78.75% (17 lines uncovered)

**Required Action**: Improve test coverage for error handling paths and edge cases in these critical components to meet repository standards (typically 85%+).

### 3. Required Workflow Check Status
The status checks show that all required workflows have passed, including:
- All security scans (Bandit, CodeQL, Semgrep) ✅
- All build processes ✅
- All pre-commit checks ✅
- fail-by-label workflow ✅

However, the "Perform Code Review" workflow is still pending, which this review addresses.

## Positive Aspects

- **Architecture Excellence**: Clean separation of concerns with dedicated handlers for different aspects (EKS, K8s, CloudWatch, IAM)
- **Security Compliance**: All security scans pass, demonstrating good security practices
- **Documentation Quality**: Comprehensive README with detailed setup instructions and usage examples
- **Standards Compliance**: Follows established MCP server patterns and naming conventions
- **Infrastructure Design**: Proper use of CloudFormation for reliable, repeatable deployments
- **Comprehensive Functionality**: Covers the full lifecycle of EKS cluster management

## Technical Review Findings

1. **Code Structure**: Well-organized with appropriate use of handlers, models, and utilities
2. **Error Handling**: Consistent error handling patterns throughout
3. **Authentication**: Proper AWS credential management and IAM integration
4. **Security Features**: Configurable read-only mode and sensitive data access controls
5. **Logging**: Consistent use of structured logging with request IDs

## Action Items

1. **HIGH PRIORITY**: Improve test coverage for core handler components:
   - Add comprehensive error handling tests
   - Include edge case scenarios
   - Test async operations thoroughly

2. **RECOMMENDED**: Address any remaining Codecov feedback to improve overall coverage

3. **VERIFICATION**: Ensure all JSON configuration examples remain valid (they appear correct in current state)

## Conclusion

This is a high-quality, comprehensive addition to the MCP ecosystem that will provide significant value for EKS users. The implementation demonstrates good engineering practices and follows repository standards. The primary blocker is the test coverage gap in core components.

Once the test coverage is improved to meet repository standards, this PR will be ready for approval.

**Decision**: REQUEST_CHANGES

**Primary Reason**: Test coverage below repository standards for core components

Reference workflow run: 15313226341

🤖 Generated with [Claude Code](https://claude.ai/code)

Co-Authored-By: Claude <noreply@anthropic.com>
EOF < /dev/null

**Review by Shubhranshu153 — APPROVED** [review]
